parsimonious/expressions.py

```diff
@@ -169,13 +167,15 @@ def match_core(self, text, pos, cache, error):
         # only the results of entire rules, not subexpressions (probably a
         # horrible idea for rules that need to backtrack internally a lot). (2)
         # Age stuff out of the cache somehow. LRU? (3) Cuts.
-        expr_id = id(self)
-        node = cache.get((expr_id, pos), MARKER)  # TODO: Change to setdefault to prevent infinite recursion in left-recursive rules.
-        if node is MARKER:
-            node = cache[(expr_id, pos)] = self._uncached_match(text,
-                                                                pos,
-                                                                cache,
-                                                                error)
+        expr_cache = cache[id(self)]
+        if pos in expr_cache:
```

Comment:
Fascinating that this makes it so much faster! I wonder if it would be faster yet using .get() so we don't do 2 separate hash lookups (one for `in` and one for `expr_cache[pos]`).

Reply:
I've benchmarked code with `.get()`:
```python
...
expr_cache = cache[id(self)]
node = expr_cache.get(pos, MARKER)
if node is MARKER:
    node = expr_cache[pos] = self._uncached_match(text,
                                                  pos,
                                                  cache,
                                                  error)
...
```
and this modification results in small slowdown:
```
+-----------+--------+----------------------------+
| Benchmark | in     | get                        |
+===========+========+============================+
| timeit    | 111 ms | 118 ms: 1.06x slower (+6%) |
+-----------+--------+----------------------------+
```
I think this happens because most of the time there is no cache entry for current parsing position, so second lookup doesn't performed too often. A trace with line execution count shows this:
```python
36359:         expr_cache = cache[id(self)]
36359:         if pos in expr_cache:
 2738:             node = expr_cache[pos]
               else:
                   # TODO: Set default value to prevent infinite recursion in left-recursive rules.
33621:             node = expr_cache[pos] = self._uncached_match(text,
```

Reply:
Also fascinating. You never know (especially with CPython) until you bench. Thanks for trying it!